<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
<meta name="viewport"
      content="width=device-width, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
<meta http-equiv="X-UA-Compatible" content="ie=edge">

    <meta name="author" content="Zachaxy">





<title>手写JVM系列(18)异常处理机制 | Hexo</title>



    <link rel="icon" href="/image/favicon.ico">




    <!-- stylesheets list from _config.yml -->
    
    <link rel="stylesheet" href="/css/style.css">
    



    <!-- scripts list from _config.yml -->
    
    <script src="/js/script.js"></script>
    
    <script src="/js/tocbot.min.js"></script>
    



    
    
        
    


<meta name="generator" content="Hexo 6.3.0"></head>

<body>
    <script>
        // this function is used to check current theme before page loaded.
        (() => {
            const currentTheme = window.localStorage && window.localStorage.getItem('theme') || '';
            const isDark = currentTheme === 'dark';
            const pagebody = document.getElementsByTagName('body')[0]
            if (isDark) {
                pagebody.classList.add('dark-theme');
                // mobile
                document.getElementById("mobile-toggle-theme").innerText = "· Dark"
            } else {
                pagebody.classList.remove('dark-theme');
                // mobile
                document.getElementById("mobile-toggle-theme").innerText = "· Light"
            }
        })();
    </script>

    <div class="wrapper">
        <header>
    <nav class="navbar">
        <div class="container">
            <div class="navbar-header header-logo"><a href="/">Zachaxy&#39;s Blog</a></div>
            <div class="menu navbar-right">
                
                    <a class="menu-item" href="/archives">Posts</a>
                
                    <a class="menu-item" href="/category">Categories</a>
                
                    <a class="menu-item" href="/tag">Tags</a>
                
                    <a class="menu-item" href="/about">About</a>
                
                <input id="switch_default" type="checkbox" class="switch_default">
                <label for="switch_default" class="toggleBtn"></label>
            </div>
        </div>
    </nav>

    
    <nav class="navbar-mobile" id="nav-mobile">
        <div class="container">
            <div class="navbar-header">
                <div>
                    <a href="/">Zachaxy&#39;s Blog</a><a id="mobile-toggle-theme">·&nbsp;Light</a>
                </div>
                <div class="menu-toggle" onclick="mobileBtn()">&#9776; Menu</div>
            </div>
            <div class="menu" id="mobile-menu">
                
                    <a class="menu-item" href="/archives">Posts</a>
                
                    <a class="menu-item" href="/category">Categories</a>
                
                    <a class="menu-item" href="/tag">Tags</a>
                
                    <a class="menu-item" href="/about">About</a>
                
            </div>
        </div>
    </nav>

</header>
<script>
    var mobileBtn = function f() {
        var toggleMenu = document.getElementsByClassName("menu-toggle")[0];
        var mobileMenu = document.getElementById("mobile-menu");
        if(toggleMenu.classList.contains("active")){
           toggleMenu.classList.remove("active")
            mobileMenu.classList.remove("active")
        }else{
            toggleMenu.classList.add("active")
            mobileMenu.classList.add("active")
        }
    }
</script>
            <div class="main">
                <div class="container">
    
    
        <div class="post-toc">
    <div class="tocbot-list">
    </div>
    <div class="tocbot-list-menu">
        <a class="tocbot-toc-expand" onclick="expand_toc()">Expand all</a>
        <a onclick="go_top()">Back to top</a>
        <a onclick="go_bottom()">Go to bottom</a>
    </div>
</div>

<script>
    var tocbot_timer;
    var DEPTH_MAX = 6; // 为 6 时展开所有
    var tocbot_default_config = {
        tocSelector: '.tocbot-list',
        contentSelector: '.post-content',
        headingSelector: 'h1, h2, h3, h4, h5',
        orderedList: false,
        scrollSmooth: true,
        onClick: extend_click,
    };

    function extend_click() {
        clearTimeout(tocbot_timer);
        tocbot_timer = setTimeout(function() {
            tocbot.refresh(obj_merge(tocbot_default_config, {
                hasInnerContainers: true
            }));
        }, 420); // 这个值是由 tocbot 源码里定义的 scrollSmoothDuration 得来的
    }

    document.ready(function() {
        tocbot.init(obj_merge(tocbot_default_config, {
            collapseDepth: 1
        }));
    });

    function expand_toc() {
        var b = document.querySelector('.tocbot-toc-expand');
        var expanded = b.getAttribute('data-expanded');
        expanded ? b.removeAttribute('data-expanded') : b.setAttribute('data-expanded', true);
        tocbot.refresh(obj_merge(tocbot_default_config, {
            collapseDepth: expanded ? 1 : DEPTH_MAX
        }));
        b.innerText = expanded ? 'Expand all' : 'Collapse all';
    }

    function go_top() {
        window.scrollTo(0, 0);
    }

    function go_bottom() {
        window.scrollTo(0, document.body.scrollHeight);
    }

    function obj_merge(target, source) {
        for (var item in source) {
            if (source.hasOwnProperty(item)) {
                target[item] = source[item];
            }
        }
        return target;
    }
</script>
    

    
    <article class="post-wrap">
        <header class="post-header">
            <h1 class="post-title">手写JVM系列(18)异常处理机制</h1>
            
                <div class="post-meta">
                    
                        Author: <a itemprop="author" rel="author" href="/">Zachaxy&nbsp;&#124;&nbsp;</a>
                    

                    
                        <span class="post-time">
                        Date: <a href="#">January 20, 2018&nbsp;&nbsp;16:58:41&nbsp;&#124;&nbsp;</a>
                        </span>
                    

                    
                        <span class="post-count">
                        Words: <a href="#">5.9k&nbsp;&#124;&nbsp;</a>
                        </span>
                    

                    
                        <span class="post-readtime">
                        About: <a href="#">5 mins.</a>
                        </span>
                    

                    
                </div>
            
        </header>

        <div class="post-content">
            <p>前面介绍了方法的调用过程，简单的方法可以在我们自己写的 JVM 中运行起来了，但是方法的执行过程并不完整，因为方法可能会抛出异常，我们暂时还无法处理异常。本节依然通过示例来讲解异常的处理机制。进一步完善我们的 JVM。核心是 <a target="_blank" rel="noopener" href="https://github.com/zachaxy/JVM/blob/master/Java/src/instructions/references/ATHROW.java">athrow 指令的实现</a></p>
<span id="more"></span>

<h1 id="本节异常处理机制所用的示例"><a href="#本节异常处理机制所用的示例" class="headerlink" title="本节异常处理机制所用的示例"></a>本节异常处理机制所用的示例</h1><p>本节所讲解的异常处理机制是基于以下面的例子做说明的，有两个方法，(1)func 方法直接抛出了一个异常；(2)test 方法调用了 func 方法，同时试图用 try-catch 来捕获异常，但是其所捕获的异常并非 func 抛出的异常，也就是说 test 方法处理不了，最终交由 JVM 来处理，而 JVM 默认的处理方式则是打印虚拟机堆栈信息。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">test</span><span class="params">()</span> &#123;</span><br><span class="line">    <span class="type">int</span> <span class="variable">i</span> <span class="operator">=</span> <span class="number">100</span>;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        func();</span><br><span class="line">    &#125; <span class="keyword">catch</span> (NumberFormatException e) &#123;</span><br><span class="line">        i = <span class="number">200</span>;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">func</span><span class="params">()</span> &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">IndexOutOfBoundsException</span>();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>test 方法的字节码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"> 0 bipush 100</span><br><span class="line"> 2 istore_0</span><br><span class="line"> 3 invokestatic #60 &lt;test/TestException12.func&gt;</span><br><span class="line"> 6 goto 14 (+8)</span><br><span class="line"> 9 astore_1</span><br><span class="line">10 sipush 200</span><br><span class="line">13 istore_0</span><br><span class="line">14 return</span><br></pre></td></tr></table></figure>

<h1 id="直观的感受-athrow-命令"><a href="#直观的感受-athrow-命令" class="headerlink" title="直观的感受 athrow 命令"></a>直观的感受 athrow 命令</h1><p>由上面的 func 方法，其方法体直接通过 throw 关键字抛出了一个异常，那么查看 func 方法对应的字节码:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">0 new #62 &lt;java/lang/IndexOutOfBoundsException&gt;</span><br><span class="line">3 dup</span><br><span class="line">4 invokespecial #63 &lt;java/lang/IndexOutOfBoundsException.&lt;init&gt;&gt;</span><br><span class="line">7 athrow</span><br></pre></td></tr></table></figure>
<p>前面的三个指令是在创建 IndexOutOfBoundsException 异常的一个实例，并执行其 init 方法。<br>接下来就是一个 athrow 命令，将上面创建的异常抛出。这就是 athrow 指令出现的场景。现在只是只管的感受一下 athrow 指令，暂时先不去实现它。</p>
<h1 id="异常处理表"><a href="#异常处理表" class="headerlink" title="异常处理表"></a>异常处理表</h1><p>func 方法内部抛出了异常，那么接下来需要另一个方法 test 来调用 func 方法。在 test 方法中，我们 catch 了一个异常 NumberFormatException，但是该异常的类型并不是 func 方法抛出的异常，因此是不会执行对应的 catch 块的。我们通过定义一个变量 i 来标示，如果执行了 catch 块，那么该变量 i 会变为 200，否则还是初值 100；</p>
<p>如果在方法中使用了 try-catch 语句，那么在该方法的 code 属性中就会有一个异常处理表。这里之所以叫表，是因为我们可能会 catch 多个异常，每 catch 一个异常，就会产生一条表项。<br>在上面的 test 方法中就有异常表，因为我们这里只有一个 catch，所以异常表只有一条表项</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">&#123;</span><br><span class="line">    start:3</span><br><span class="line">    end:6</span><br><span class="line">    hadlerPC:9</span><br><span class="line">    catchType: #5 (java/lang/NumberFormatException)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>JVM 中定义的异常处理表项有四个字段：均为整数</p>
<ul>
<li>startPc</li>
<li>endPc</li>
<li>handlerPc</li>
<li>catchType</li>
</ul>
<p>其含义如下：<br>[startPc,endPc) 一个区间:锁定一个指令区间,该区间可能抛出某种异常。test 的异常表表示在字节码 3 到字节码 6(不包含 6)之间，可能会抛出异常，而指令 3 正是调用了抛出异常的 func 方法。<br>handlerPc:负责处理异常的 catch 块的其实位置，一旦发现某个 catch 可以处理该异常，那么就将 nextPc 改为 handlerPc。<br>catchType:指向运行时常量池的一个索引,通过该索引可以得到一个异常的类 xxxException 的对象，也就是我们在代码中显式 catch 的异常的类型。</p>
<h1 id="解析异常处理表"><a href="#解析异常处理表" class="headerlink" title="解析异常处理表"></a>解析异常处理表</h1><p>和之前的常量池类似，异常处理表也是保存 class 文件中，我们在读取 class 文件的时候，一样要对其进行转换。<br>首先看一下其在 ClassFile 中的表现形式:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">class</span> <span class="title class_">ExceptionTableEntry</span> &#123;</span><br><span class="line">    <span class="type">int</span> startPc;        <span class="comment">//可能排除异常的代码块的起始字节码（包括）</span></span><br><span class="line">    <span class="type">int</span> endPc;          <span class="comment">//可能排除异常的代码块的终止字节码（不包括）</span></span><br><span class="line">    <span class="type">int</span> handlerPc;      <span class="comment">//负责处理异常的 catch 块的其实位置</span></span><br><span class="line">    <span class="type">int</span> catchType;      <span class="comment">//指向运行时常量池的一个索引，解析后可以得到一个异常类</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>其在 Zclass 中的表现：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">ExceptionHandler</span> &#123;</span><br><span class="line">    <span class="type">int</span> startPc;</span><br><span class="line">    <span class="type">int</span> endPc;</span><br><span class="line">    <span class="type">int</span> handlerPc;</span><br><span class="line">    ClassRef catchType;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>对比发现，具体的转换是将 catchType 运行时常量池的常量直接转换为对应的异常类的引用，其转换过程是在 ClassFile-&gt;Zclass 的转换过程中，具体对应到方法的转换过程。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="title function_">Zclass</span><span class="params">(ClassFile classFile)</span> &#123;</span><br><span class="line">    ...</span><br><span class="line">    runtimeConstantPool = <span class="keyword">new</span> <span class="title class_">RuntimeConstantPool</span>(<span class="built_in">this</span>, classFile.getConstantPool());</span><br><span class="line">    fileds = Zfield.makeFields(<span class="built_in">this</span>, classFile.getFields());</span><br><span class="line">    methods = Zmethod.makeMethods(<span class="built_in">this</span>, classFile.getMethods());</span><br><span class="line">    ...</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>



<h1 id="athorw-命令的实现-异常处理流程"><a href="#athorw-命令的实现-异常处理流程" class="headerlink" title="athorw 命令的实现(异常处理流程)"></a>athorw 命令的实现(异常处理流程)</h1><p>再次贴出 test 方法的代码</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">test</span><span class="params">()</span> &#123;</span><br><span class="line">    <span class="type">int</span> <span class="variable">i</span> <span class="operator">=</span> <span class="number">100</span>;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        func();</span><br><span class="line">    &#125; <span class="keyword">catch</span> (NumberFormatException e) &#123;</span><br><span class="line">        i = <span class="number">200</span>;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>test 对应的栈帧为 frame1，当其内部执行 func 方法时，会为其创建栈帧 frame2，在 frame2 中执行 func 方法，在 func 方法中，首先创建了一个异常——IndexOutOfBoundsException 的对象，该对象放在 frame2 的操作数栈顶。然后遇到了 athrow 命令，该指令将 frame2 操作数栈顶的异常对象 ex pop 出来，然后从当前 frame 开始遍历虚拟机栈。我们知道当前 JVM 的栈为 frame1 -&gt; frame2。那么先从 frame2 开始，查找当前方法的异常处理表，当然我们一般在写代码的时候不会在手动写了 throw 异常之后，立刻在当前方法就去 catch 它，我们一般的做法都是在 func 方法的声明处显示的 throws 对应的异常，等待调用 func 的上层方法来出来，但是语法也是允许自己来 catch 的，所以先从 frame2 查找，查找不到，那么将 frame2 从虚拟机栈弹出，然后再从当前栈顶的 frame1 中查找是否可以找到异常处理项，就这样一直遍历下去。</p>
<p>这样的遍历会产生两个结果：</p>
<ol>
<li>在某个 frame 中找到了可以处理该异常的项，那么首先到该 frame 的操作数栈清空，然后把异常对象压入当前操作数栈，接着执行该 frame 对应的 catch 块语句。</li>
<li>遍历了整个虚拟机栈都没有找到可以处理的异常项，因此交由 JVM 来处理，而 JVM 的处理方式就是打印虚拟机栈信息(从产生异常的方法 func 对应的 frame，依次到 main 方法对应的 frame)</li>
</ol>
<p>athrow 指令的处理流程：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="type">Zobject</span> <span class="variable">ex</span> <span class="operator">=</span> frame.getOperandStack().popRef();</span><br><span class="line"></span><br><span class="line"><span class="comment">//获取当前 thread,是想从 thread 中获取所有的 frame(一个 frame 就是一个方法调用过程的抽象)</span></span><br><span class="line"><span class="type">Zthread</span> <span class="variable">thread</span> <span class="operator">=</span> frame.getThread();</span><br><span class="line"><span class="keyword">if</span> (!findAndGotoExceptionHandler(thread, ex)) &#123;</span><br><span class="line">    <span class="comment">//所有的方法调用栈均不能处理该异常了,那么交由 JVM 来处理;</span></span><br><span class="line">    handleUncaughtException(thread, ex);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>上面多次提到了在当前 frame 寻找对应的异常处理项，如何寻找，这就用到了 findAndGotoExceptionHandler 方法，具体流程如下：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"> <span class="comment">//从当前帧开始,遍历 Java 虚拟机栈,查找方法的异常处理表</span></span><br><span class="line"><span class="keyword">private</span> <span class="type">boolean</span> <span class="title function_">findAndGotoExceptionHandler</span><span class="params">(Zthread thread, Zobject ex)</span> &#123;</span><br><span class="line">    <span class="keyword">while</span> (<span class="literal">true</span>) &#123;</span><br><span class="line">        <span class="type">Zframe</span> <span class="variable">frame</span> <span class="operator">=</span> thread.getCurrentFrame();</span><br><span class="line">        <span class="comment">//正常情况下,获取一条指令后,bytereader 中的 pc 是要+1,指向下一条指令的地址</span></span><br><span class="line">        <span class="comment">// 但是 athrow 指令比较特殊,因为现在已经抛出异常了,不能继续向下执行了,所以要回退;</span></span><br><span class="line">        <span class="type">int</span> <span class="variable">pc</span> <span class="operator">=</span> frame.getNextPC() - <span class="number">1</span>;</span><br><span class="line"></span><br><span class="line">        <span class="comment">//此时 pc 指向的是 throw 异常的那一行;因为接下来就要在当前方法的异常处理表中寻找可以处理当前 pc 指向的指令所抛出的异常了</span></span><br><span class="line">        <span class="type">int</span> <span class="variable">handlerPC</span> <span class="operator">=</span> frame.getMethod().findExceptionHandler(ex.getClazz(), pc);</span><br><span class="line"></span><br><span class="line">        <span class="comment">//当前方法能处理自己抛出的异常</span></span><br><span class="line">        <span class="keyword">if</span> (handlerPC &gt; <span class="number">0</span>) &#123;</span><br><span class="line">            <span class="type">OperandStack</span> <span class="variable">operandStack</span> <span class="operator">=</span> frame.getOperandStack();</span><br><span class="line">            operandStack.clear();<span class="comment">//清空操作数栈</span></span><br><span class="line">            operandStack.pushRef(ex);<span class="comment">//将抛出的异常放进去</span></span><br><span class="line">            frame.setNextPC(handlerPC);<span class="comment">//指令跳转到对应的 catch 块中</span></span><br><span class="line">            <span class="keyword">return</span> <span class="literal">true</span>;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        <span class="comment">//能走到这一步,说明当前方法不能处理当前方法抛出的异常,需要回到调用该方法的帧 frame</span></span><br><span class="line">        thread.popFrame();</span><br><span class="line"></span><br><span class="line">        <span class="comment">//整个调用栈都无法处理异常，交由 JVM 来处理吧；JVM 处理的方法就是下面的 handleUncaughtException</span></span><br><span class="line">        <span class="keyword">if</span> (thread.isStackEmpty()) &#123;</span><br><span class="line">            <span class="keyword">break</span>;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">false</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>超找异常处理项，最终是在 Method 的 ExceptionTable 对象中</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">//返回能解决当前 Exception 的 handler=&gt;多个 catch 块,决定用哪个</span></span><br><span class="line"><span class="keyword">public</span> ExceptionHandler <span class="title function_">findExceptionHandler</span><span class="params">(Zclass exClazz, <span class="type">int</span> pc)</span> &#123;</span><br><span class="line">    <span class="keyword">for</span> (<span class="type">int</span> <span class="variable">i</span> <span class="operator">=</span> <span class="number">0</span>; i &lt; exceptionTable.length; i++) &#123;</span><br><span class="line">        <span class="type">ExceptionHandler</span> <span class="variable">handler</span> <span class="operator">=</span> exceptionTable[i];</span><br><span class="line">        <span class="keyword">if</span> (pc &gt;= handler.startPc &amp;&amp; pc &lt; handler.endPc) &#123;</span><br><span class="line">            <span class="comment">// catch all</span></span><br><span class="line">            <span class="keyword">if</span> (handler.catchType == <span class="literal">null</span>) &#123;</span><br><span class="line">                <span class="keyword">return</span> handler;</span><br><span class="line">            &#125;</span><br><span class="line">            <span class="comment">// 如果 catch 的异常是实际抛出的异常的父类，也可以捕获</span></span><br><span class="line">            <span class="type">Zclass</span> <span class="variable">catchClazz</span> <span class="operator">=</span> handler.catchType.resolvedClass();</span><br><span class="line">            <span class="keyword">if</span> (catchClazz == exClazz || catchClazz.isSuperClassOf(exClazz)) &#123;</span><br><span class="line">                <span class="keyword">return</span> handler;</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">null</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>上面能找到对应的异常处理项，那么整个程序还可以正常运行，但如果没有找到的话，就要交给 JVM 来处理的，此时也意味着当前的虚拟机栈已经被清空了，解释器也停止执行了，JVM 将整个调用链打印出来，程序结束。</p>
<h1 id="Java-虚拟机栈信息"><a href="#Java-虚拟机栈信息" class="headerlink" title="Java 虚拟机栈信息"></a>Java 虚拟机栈信息</h1><figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">NStackTraceElement</span> &#123;</span><br><span class="line">    String fileName;<span class="comment">//类所在的 java 文件</span></span><br><span class="line">    String className;<span class="comment">//声明方法的类名</span></span><br><span class="line">    String methodName;<span class="comment">//调用方法名</span></span><br><span class="line">    <span class="type">int</span> lineNumber;<span class="comment">//出现 exception 的行号</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>NStackTraceElement 代表每个 frame 中调用抛出异常方法的具体消息。</p>
<p>上面实现了 ATHROW 命令，但是目前程序还是无法运行，因为异常的父类 Exception 的构造方法中会调用一个 native 的方法——fillInStackTrace，因此还要在 native 方法中注册 fillInStackTrace 方法，同时在该方法中进行虚拟机调用栈的收集工作。<br>具体的代码请参考<a target="_blank" rel="noopener" href="https://github.com/zachaxy/JVM/blob/master/Java/src/znative/java/lang/Nthrowable">Nthrowable.java</a>，注释已经写的很清楚，相信不会有太大的阅读困难。</p>
<h1 id="异常的测试用例"><a href="#异常的测试用例" class="headerlink" title="异常的测试用例"></a>异常的测试用例</h1><p>同样，这里提供了一个单独的<a target="_blank" rel="noopener" href="https://github.com/zachaxy/JVM/blob/master/Java/src/test/TestException12.java">测试用例</a>来测试异常，我们的 test 方法并没有不会 func 方法抛出的异常，这一点从最终打印 test 的本地变量表就可以发现，i 的值为 100，没有走到 catch 语句中。同时程序也打印出了虚拟机堆栈信息，具体打印信息如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">the same as testInterpreter06!</span><br><span class="line">java -cp /Users/zachaxy/TestClassFiles  TestException12</span><br><span class="line">current instruction: 0: BIPUSH</span><br><span class="line">current instruction: 2: ISTORE_0</span><br><span class="line">current instruction: 3: INVOKE_STATIC</span><br><span class="line">current instruction: 3: INVOKE_STATIC</span><br><span class="line">test.TestException12.f1(TestException12.java:82)</span><br><span class="line">test.TestException12.test(TestException12.java:75)</span><br><span class="line">100</span><br></pre></td></tr></table></figure>
        </div>

        
            <section class="post-copyright">
                
                    <p class="copyright-item">
                        <span>本文作者:</span>
                        <span>Zachaxy</span>
                    </p>
                
                
                
                    <p class="copyright-item">
                        <span>版权声明:</span>
                        <span>著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。</span>
                    </p>
                
                

            </section>
        
        <section class="post-tags">
            <div>
                <span>Tag(s):</span>
                <span class="tag">
                    
                    
                        <a href="/tags/JVM/"># JVM</a>
                    
                        
                </span>
            </div>
            <div>
                <a href="javascript:window.history.back();">back</a>
                <span>· </span>
                <a href="/">home</a>
            </div>
        </section>
        <section class="post-nav">
            
                <a class="prev" rel="prev" href="/2018/01/23/TestMusic/">TestMusic</a>
            
            
            <a class="next" rel="next" href="/2018/01/18/%E6%89%8B%E5%86%99JVM%E7%B3%BB%E5%88%97-17-%E6%9C%AC%E5%9C%B0%E6%96%B9%E6%B3%95%E7%9A%84%E8%B0%83%E7%94%A8%E8%BF%87%E7%A8%8B/">手写JVM系列(17)本地方法的调用过程</a>
            
        </section>


    </article>
</div>

            </div>
            <footer id="footer" class="footer">
    <div class="copyright">
        <span>© Zachaxy | Powered by <a href="https://hexo.io" target="_blank">Hexo</a> & <a href="https://github.com/Siricee/hexo-theme-Chic" target="_blank">Chic</a></span>
    </div>
</footer>

    </div>
</body>

</html>